platform/cog-platform-gtk4.c

```diff
@@ -0,0 +1,651 @@
+}
+
+static void on_forward_clicked(GtkWidget* widget, gpointer user_data)
+{
+    webkit_web_view_go_forward(win.web_view);
+}
+
+static void on_refresh_clicked(GtkWidget* widget, gpointer user_data)
+{
+    webkit_web_view_reload(win.web_view);
+}
+
+static void on_entry_activated(GtkWidget* widget, gpointer user_data)
+{
+    GtkEntryBuffer* buffer = gtk_entry_get_buffer(GTK_ENTRY(win.url_entry));
+    const gchar* uri = gtk_entry_buffer_get_text(buffer);
+    webkit_web_view_load_uri(win.web_view, uri);
+}
+
+static gboolean action_quit(GtkWidget* widget, GVariant* args,
+    gpointer user_data)
+{
+    if (g_application_get_default())
+        g_application_quit(g_application_get_default());
+
+    return TRUE;
+}
+
+static gboolean action_activate_entry(GtkWidget* widget, GVariant* args,
+    gpointer user_data)
+{
+    gtk_widget_grab_focus(win.url_entry);
+    return TRUE;
+}
+
+static void setup_window(struct platform_window* window)
+{
+    g_assert_nonnull(window);
+    window->gtk_window = gtk_window_new();
+    g_object_set(window->gtk_window, "default-width", DEFAULT_WIDTH,
+        "default-height", DEFAULT_HEIGHT, NULL);
+    g_signal_connect(window->gtk_window, "destroy", G_CALLBACK(quit_cb), NULL);
+
+    GtkWidget* header_bar = gtk_header_bar_new();
+    GtkWidget* left_stack = gtk_stack_new();
+    GtkWidget* buttons_box = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 0);
+
+    window->back_button = gtk_button_new_from_icon_name("go-previous-symbolic");
+    g_signal_connect(window->back_button, "clicked", G_CALLBACK(on_back_clicked),
+        NULL);
+    gtk_box_append(GTK_BOX(buttons_box), window->back_button);
+
+    window->forward_button = gtk_button_new_from_icon_name("go-next-symbolic");
+    g_signal_connect(window->forward_button, "clicked",
+        G_CALLBACK(on_forward_clicked), NULL);
+    gtk_box_append(GTK_BOX(buttons_box), window->forward_button);
+
+    GtkWidget* refresh_button = gtk_button_new_from_icon_name("view-refresh-symbolic");
+    g_signal_connect(refresh_button, "clicked", G_CALLBACK(on_refresh_clicked),
+        NULL);
+    gtk_box_append(GTK_BOX(buttons_box), refresh_button);
+
+    gtk_stack_add_named(GTK_STACK(left_stack), buttons_box, "buttons");
+    gtk_header_bar_pack_start(GTK_HEADER_BAR(header_bar), left_stack);
+
+    GtkWidget* title_box = gtk_box_new(GTK_ORIENTATION_VERTICAL, 0);
+    gtk_widget_set_valign(title_box, GTK_ALIGN_CENTER);
+    gtk_widget_set_hexpand(title_box, TRUE);
+
+    window->url_entry = gtk_entry_new();
+    g_signal_connect(window->url_entry, "activate",
+        G_CALLBACK(on_entry_activated), NULL);
+    gtk_widget_set_parent(window->url_entry, title_box);
+
+    gtk_header_bar_set_title_widget(GTK_HEADER_BAR(header_bar), title_box);
+    gtk_header_bar_set_show_title_buttons(GTK_HEADER_BAR(header_bar), TRUE);
+    gtk_window_set_titlebar(GTK_WINDOW(window->gtk_window), header_bar);
+
+    window->gl_drawing_area = gtk_gl_area_new();
```

Comment:
As a future improvement idea, I suppose that instead of using `GtkGLArea` maybe it could be possible to implement a `GtkWidget` which would use [gtk_snapshot_append_texture()](https://developer.gnome.org/gtk4/4.0/GtkSnapshot.html#gtk-snapshot-append-texture) to directly pass the GL texture, and let GTK do the compositing—we would not need to provide shaders to scan out the texture when painting a quad using GL calls, and if I understand correctly it should be faster because `GtkGLArea` paints to a FBO backed by a texture, and then _that_ texture is snapshotted by the `GtkGLArea widget implementation.

Reply:
Filed as #276